### PR TITLE
Fix VISSL on GPU and add VISSL GPU CI

### PR DESCRIPTION
**Pull request by ethanwharris**
## What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the **[contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md)**, Pull Request section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes?
- [x] Did you write any **new necessary tests**? [not needed for typos/docs]
- [ ] Did you verify **new and existing tests pass** locally with your changes?
- [ ] If you made a notable change (that affects users), did you **update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)**?

## PR review
 - [x] Is this pull request **ready for review**? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/reference/image_embedder.rst [hunk 1]
```diff
@@ -4,6 +4,10 @@
    :image: https://pl-flash-data.s3.amazonaws.com/assets/thumbnails/image_embedder.svg
    :tags: Image,Embedding
 
+.. warning::
+
+   Multi-gpu training is not currently supported by the :class:`~flash.image.embedding.model.ImageEmbedder` task.
+
```
Comment (krshrimali) [inline docs/source/reference/image_embedder.rst:10]: Let's create an issue for this to track? :)

File: docs/source/reference/image_embedder.rst [hunk 2]
```diff
@@ -26,18 +32,29 @@ Example
...
+The ``head`` and ``pretraining_transform`` arguments should match the choice of ``training_strategy`` following this table:
+
+=====================  ================  ==========================
+``training_strategy``  ``head``          ``pretraining_transform``
+=====================  ================  ==========================
+``simclr``             ``simclr_head``   ``simclr_transform``
+``barlow_twins``       ``barlow_twins``  ``barlow_twins_transform``
+``swav``               ``swav_head``     ``swav_transform``
+``dino``               ``dino_head``     ``dino_transform``
+=====================  ================  ==========================
```
Comment (krshrimali) [inline docs/source/reference/image_embedder.rst:60]: Excellent!!!

File: flash/image/embedding/vissl/hooks.py [hunk 3]
```diff
@@ -81,6 +82,10 @@ def on_start(self, task: "flash.image.embedding.vissl.adapter.MockVISSLTask") ->
 
         task.loss.info_criterion.precompute_pos_neg_mask()
 
+        # Re-register params / devices
+        _recursive_register(task.loss)
+        task.loss.to(task.vissl_adapter.adapter_task.device, task.vissl_adapter.adapter_task.dtype)
```
Comment (krshrimali) [inline flash/image/embedding/vissl/hooks.py:87]: Do we need to register again? Wouldn't just `task.loss.to(<device>, <dtype>)` work?

Checked locally: tests pass if we remove `_recursive_register` here. But I maybe wrong. :)

File: flash_examples/image_embedder.py [hunk 4]
```diff
@@ -21,23 +21,21 @@
 # 1. Download the data and prepare the datamodule
 datamodule = ImageClassificationData.from_datasets(
     train_dataset=CIFAR10(".", download=True),
-    batch_size=16,
+    batch_size=4,
```
Comment (krshrimali) [inline flash_examples/image_embedder.py:24]: Less batch size, more steps, right? Does this make CI slower?

**Review by krshrimali — APPROVED** [review]
Thanks, @ethanwharris - just a couple of comments, but overall looks great. 🔥 🚀

File: flash_examples/image_embedder.py (hunk 4 above)
Comment (ethanwharris) [reply flash_examples/image_embedder.py:24]: I think the GPU CI internally overwrites the batch size to be higher, but thinks makes things quicker on the CPU CI (and consistent with what we have in other places) :smiley:

File: flash/image/embedding/vissl/hooks.py (hunk 3 above)
Comment (ethanwharris) [reply flash/image/embedding/vissl/hooks.py]: Removed :smiley:

File: flash/image/embedding/vissl/hooks.py [hunk 5]
```diff
@@ -81,6 +81,9 @@ def on_start(self, task: "flash.image.embedding.vissl.adapter.MockVISSLTask") ->
 
         task.loss.info_criterion.precompute_pos_neg_mask()
 
+        # Cast the loss to the correct device / dtype
+        task.loss.to(task.vissl_adapter.adapter_task.device, task.vissl_adapter.adapter_task.dtype)
```
Comment (krshrimali) [inline flash/image/embedding/vissl/hooks.py:85]: minor nit :)

```suggestion
        task.loss.to(lightning_module.device, lightning_module.dtype)
```

**Review by Borda — APPROVED** [review]

File: docs/source/reference/image_embedder.rst (hunk 1 above)
Comment (ethanwharris) [reply docs/source/reference/image_embedder.rst:10]: I think it's not possible to fix this for VISSL so I'll leave it for now
